compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4664,24 +4669,33 @@ private void visitCheckAndCheckPanicExpr(BLangCheckedExpr checkedExpr) {
             return;
         }
 
-        BUnionType unionType = (BUnionType) exprType;
         // Filter out the list of types which are not equivalent with the error type.
-        Map<Boolean, List<BType>> resultTypeMap = unionType.getMemberTypes().stream()
-                .collect(Collectors.groupingBy(memberType -> types.isAssignable(memberType, symTable.errorType)));
+        List<BType> errorAssignableTypes = new ArrayList<>();
+        List<BType> errorNotAssignableTypes = new ArrayList<>();
+        for (BType memberType : ((BUnionType) exprType).getMemberTypes()) {
+            if (memberType.tag == TypeTags.READONLY) {
+                errorAssignableTypes.add(symTable.errorType);
+                errorNotAssignableTypes.add(symTable.anyAndReadonly);
+                continue;
+            }
+            if (types.isAssignable(memberType, symTable.errorType)) {
+                errorAssignableTypes.add(memberType);
+            } else {
+                errorNotAssignableTypes.add(memberType);
+            }
+        }
 
         // This list will be used in the desugar phase
-        checkedExpr.equivalentErrorTypeList = resultTypeMap.get(true);
-        if (checkedExpr.equivalentErrorTypeList == null ||
-                checkedExpr.equivalentErrorTypeList.size() == 0) {
+        checkedExpr.equivalentErrorTypeList = errorAssignableTypes;
+        if (checkedExpr.equivalentErrorTypeList.size() == 0) {
```

Comment:
We can directly check
```suggestion
        if (errorAssignableTypes.isEmpty()) {
```
right?

Reply:
Changed.